keps/sig-network/0015-20180614-SCTP-support.md

```diff
@@ -0,0 +1,192 @@
+    targetPort: 9376
+```
+
+#### Headless Service with SCTP
+As a user of Kubernetes I want to define headless Services for my applications that use SCTP as L4 protocol on their interfaces, so client applications can discover my applications in kube-dns, or via any other service discovery method that gets information about endpoints via the Kubernetes API.
+Example:
+```
+kind: Service
+apiVersion: v1
+metadata:
+  name: my-service
+spec:
+  selector:
+    app: MyApp
+  ClusterIP: "None"
+  ports:
+  - protocol: SCTP
+    port: 80
+    targetPort: 9376
+```
+#### Service with SCTP without selector
+As a user of Kubernetes I want to define Services without selector for my applications that use SCTP as L4 protocol on their interfaces,so I can implement my own service controllers if I want to extend the basic functionality of Kubernetes.
+Example:
+```
+kind: Service
+apiVersion: v1
+metadata:
+  name: my-service
+spec:
+  ClusterIP: "None"
+  ports:
+  - protocol: SCTP
+    port: 80
+    targetPort: 9376
+```
+#### NetworkPolicy with SCTP
+As a user of Kubernetes I want to define NetworPolicies for my applications that use SCTP as L4 protocol on their interfaces, so the network controllers that support SCTP can control the accessibility of my applications on the SCTP based interfaces, too.
+Example:
+```
+apiVersion: networking.k8s.io/v1
+kind: NetworkPolicy
+metadata:
+  name: myservice-network-policy
+  namespace: myapp
+spec:
+  podSelector:
+    matchLabels:
+      role: myservice
+  policyTypes:
+  - Ingress
+  - Egress
+  ingress:
+  - from:
+    - ipBlock:
+        cidr: 172.17.0.0/16
+        except:
+        - 172.17.1.0/24
+    - namespaceSelector:
+        matchLabels:
+          project: myproject
+    - podSelector:
+        matchLabels:
+          role: myclient
+    ports:
+    - protocol: SCTP
+      port: 7777
+```
+#### User space SCTP stack
+As a user of Kubernetes I want to deploy and run my applications that use a user space SCTP stack. 
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### SCTP in Services
+The Kubernetes API modification for Services is obvious.
+The selected port shall be reserved on the node, just like for TCP and UDP now. Unfortunately, golang does not have native SCTP support in the "net" package, so in order to access the kernel's SCTP API we have to introduce a new 3rd party package as a new vendor package.
+For Services with type=LoadBalancer we have to check how the cloud provider implementations handle new protocols, and we have to make sure that if SCTP is not supported then the request for a new load balancer, firewall rule, etc. with protocol=SCTP is rejected gracefully.
+DNS shall support SRV records with "_sctp" as "proto" value.
+
+#### SCTP in NetworkPolicy
+The Kubernetes API modification for the NetworkPolicy is obvious.
+In order to utilize the new protocol value the network controller must support it.
+
+#### Interworking with applications that use a user space SCTP stack
```

Comment:
This feels like a separate KEP ? to enable use of user-space protocol stacks + kube-proxy.